components/proxy/src/test/kotlin/com/hotels/styx/routing/handlers/LoadBalancingGroupTest.kt

```diff
@@ -0,0 +1,218 @@
+import com.hotels.styx.api.HttpRequest.get
+import com.hotels.styx.api.configuration.ObjectStore
+import com.hotels.styx.api.exceptions.NoAvailableHostsException
+import com.hotels.styx.routing.RoutingContext
+import com.hotels.styx.routing.RoutingObjectRecord
+import com.hotels.styx.routing.db.StyxObjectStore
+import com.hotels.styx.routing.handle
+import com.hotels.styx.routing.routingObjectDef
+import com.hotels.styx.server.HttpInterceptorContext
+import io.kotlintest.IsolationMode
+import io.kotlintest.matchers.numerics.shouldBeGreaterThan
+import io.kotlintest.matchers.types.shouldBeNull
+import io.kotlintest.shouldBe
+import io.kotlintest.shouldThrow
+import io.kotlintest.specs.FeatureSpec
+import io.mockk.mockk
+import io.mockk.verify
+import org.reactivestreams.Publisher
+import reactor.core.Disposable
+import reactor.core.publisher.toFlux
+import reactor.core.publisher.toMono
+import java.nio.charset.StandardCharsets.UTF_8
+import java.time.Duration
+
+class LoadBalancingGroupTest : FeatureSpec() {
+
+    // Tests depend on each other. So run the tests sequentially:
+    override fun isolationMode(): IsolationMode = IsolationMode.SingleInstance
+
+    init {
+        feature("Load Balancing") {
+            val factory = LoadBalancingGroup.Factory()
+            val routeDb = StyxObjectStore<RoutingObjectRecord>()
+
+            routeDb.insert("appx-01", RoutingObjectRecord.create("HostProxy", setOf("appX"), mockk(), StaticResponseHandler(200, "appx-01")))
+            routeDb.insert("appx-02", RoutingObjectRecord.create("HostProxy", setOf("appX"), mockk(), StaticResponseHandler(200, "appx-02")))
+            routeDb.insert("appx-03", RoutingObjectRecord.create("HostProxy", setOf("appX"), mockk(), StaticResponseHandler(200, "appx-03")))
+
+            routeDb.insert("appy-01", RoutingObjectRecord.create("HostProxy", setOf("appY"), mockk(), StaticResponseHandler(200, "appy-01")))
+            routeDb.insert("appy-02", RoutingObjectRecord.create("HostProxy", setOf("appY"), mockk(), StaticResponseHandler(200, "appy-02")))
+
+            routeDb.watch().waitUntil { it.entrySet().size == 5 }
+
+            val lbGroup = factory.build(listOf("appX"), RoutingContext(routeDb = routeDb).get(), routingObjectDef("""
+                type: LoadBalancingGroup
+                config:
+                  origins: appX
+                """.trimIndent())) as LoadBalancingGroup
+
+            scenario("Discovers origins with appropriate tag") {
+                val frequencies = mutableMapOf<String, Int>()
+
+                for (i in 1..100) {
+                    lbGroup.call(get("/").build())
+                            .bodyAs(UTF_8)
+                            .let {
+                                val current = frequencies.getOrDefault(it, 0)
+                                frequencies[it] = current + 1
+                            }
+                }
+
+                println("Frequencies: " + frequencies)
```

Comment:
```suggestion
                
```
Remove debugging `println`s.

Reply:
If this information is needed to understand a test failure, it should be part of the failure message.

Reply:
Do you know if there is a way to print out this information after an assertion failed?

Reply:
Actually, I'm not sure. I know that we can do this in Java tests (using Hamcrest for example) but I don't know if there is a kotlintest equivalent. I would hope so.

Reply:
Do you have any examples how to do this in Hamcrest?